File: src/SlimShader/Chunks/Rdef/ResourceBinding.cs
using SlimShader.Chunks.Common;
using SlimShader.Util;

namespace SlimShader.Chunks.Rdef
{
	/// <summary>
	/// Roughly corresponds to the D3D11_SHADER_INPUT_BIND_DESC structure.
	/// </summary>
	public class ResourceBinding
	{
		/// <summary>
		/// Name of the shader resource.
		/// </summary>
		public string Name { get; private set; }

		/// <summary>
		/// Identifies the type of data in the resource.
		/// </summary>
		public ShaderInputType Type { get; private set; }

		/// <summary>
		/// Starting bind point.
		/// </summary>
		public uint BindPoint { get; private set; }

		/// <summary>
		/// Number of contiguous bind points for arrays.
		/// </summary>
		public uint BindCount { get; private set; }

		/// <summary>
		/// Shader input-parameter options.
		/// </summary>
		public ShaderInputFlags Flags { get; private set; }

		/// <summary>
		/// Identifies the dimensions of the bound resource.
		/// </summary>
		public ShaderResourceViewDimension Dimension { get; private set; }

		/// <summary>
		/// If the input is a texture, the return type.
		/// </summary>
		public ResourceReturnType ReturnType { get; private set; }

		/// <summary>
		/// The number of samples for a multisampled texture; otherwise 0.
		/// </summary>
		public uint NumSamples { get; private set; }

		public static ResourceBinding Parse(BytecodeReader reader, BytecodeReader resourceBindingReader)
		{
			uint nameOffset = resourceBindingReader.ReadUInt32();
			var nameReader = reader.CopyAtOffset((int) nameOffset);
			return new ResourceBinding
			{
				Name = nameReader.ReadString(),
				Type = (ShaderInputType) resourceBindingReader.ReadUInt32(),
				ReturnType = (ResourceReturnType) resourceBindingReader.ReadUInt32(),
				Dimension = (ShaderResourceViewDimension) resourceBindingReader.ReadUInt32(),
				NumSamples = resourceBindingReader.ReadUInt32(),
				BindPoint = resourceBindingReader.ReadUInt32(),
				BindCount = resourceBindingReader.ReadUInt32(),
				Flags = (ShaderInputFlags) resourceBindingReader.ReadUInt32()
			};
		}

		public override string ToString()
		{
			string returnType = ReturnType.GetDescription(Type);
			if (Flags.HasFlag(ShaderInputFlags.TextureComponent0) && Flags.HasFlag(ShaderInputFlags.TextureComponent1))
				returnType += "4";
			string typeDescription = Type.GetDescription();
			if (Flags.HasFlag(ShaderInputFlags.ComparisonSampler))
				typeDescription += "_c";
			return string.Format("// {0,-30} {1,10} {2,7} {3,11} {4,4} {5,8}",
				Name, typeDescription, returnType,
				Dimension.GetDescription(Type, ReturnType) + (Dimension.IsMultiSampled() ? NumSamples.ToString() : string.Empty),
				BindPoint, BindCount);
		}
	}
}

File: src/cpp/Source/Chunks/Rdef/ResourceBinding.cpp
#include "PCH.h"
#include "ResourceBinding.h"

using namespace std;
using namespace SlimShader;

ResourceBinding ResourceBinding::Parse(const BytecodeReader& reader, BytecodeReader& resourceBindingReader)
{
	auto nameOffset = resourceBindingReader.ReadUInt32();
	auto nameReader = reader.CopyAtOffset(nameOffset);

	ResourceBinding result;

	result._name = nameReader.ReadString();
	result._type = static_cast<ShaderInputType>(resourceBindingReader.ReadUInt32());
	result._returnType = static_cast<ResourceReturnType>(resourceBindingReader.ReadUInt32());
	result._dimension = static_cast<ShaderResourceViewDimension>(resourceBindingReader.ReadUInt32());
	result._numSamples = resourceBindingReader.ReadUInt32();
	result._bindPoint = resourceBindingReader.ReadUInt32();
	result._bindCount = resourceBindingReader.ReadUInt32();
	result._flags = static_cast<ShaderInputFlags>(resourceBindingReader.ReadUInt32());

	return result;
}

const std::string& ResourceBinding::GetName() const { return _name; }
ShaderInputType ResourceBinding::GetType() const { return _type; }
uint32_t ResourceBinding::GetBindPoint() const { return _bindPoint; }
uint32_t ResourceBinding::GetBindCount() const { return _bindCount; }
ShaderInputFlags ResourceBinding::GetFlags() const { return _flags; }
ShaderResourceViewDimension ResourceBinding::GetDimension() const { return _dimension; }
ResourceReturnType ResourceBinding::GetReturnType() const { return _returnType; }
uint32_t ResourceBinding::GetNumSamples() { return _numSamples; }

string ToString(ResourceReturnType value, ShaderInputType shaderInputType)
{
	if (value == ResourceReturnType::Mixed)
	{
		switch (shaderInputType)
		{
		case ShaderInputType::Structured:
		case ShaderInputType::UavRwStructured:
			return "struct";
		case ShaderInputType::ByteAddress:
		case ShaderInputType::UavRwByteAddress:
			return "byte";
		default:
			throw runtime_error("Shader input type '" + to_string((int) shaderInputType) + "' is not supported.");
		}
	}
	return ToString(value);
}

string ToString(ShaderResourceViewDimension value, ShaderInputType shaderInputType,
	ResourceReturnType format)
{
	switch (shaderInputType)
	{
	case ShaderInputType::ByteAddress :
	case ShaderInputType::Structured:
		return "r/o";
	case ShaderInputType::UavRwByteAddress :
	case ShaderInputType::UavRwStructured :
	case ShaderInputType::UavRwStructuredWithCounter :
	case ShaderInputType::UavRwTyped :
		return "r/w";
	default :
		return ToString(value);
	}
}

bool IsMultiSampled(ShaderResourceViewDimension value)
{
	switch (value)
	{
	case ShaderResourceViewDimension::Texture2DMultiSampled :
	case ShaderResourceViewDimension::Texture2DMultiSampledArray :
		return true;
	default :
		return false;
	}
}

ostream& SlimShader::operator<<(ostream& out, const ResourceBinding& value)
{
	string returnType = ::ToString(value._returnType, value._type);
	if (value._returnType != ResourceReturnType::NotApplicable && value._returnType != ResourceReturnType::Mixed)
		returnType += "4";
	out << boost::format("// %-30s %10s %7s %11s %4i %8i")
		% value._name
		% ToString(value._type)
		% returnType
		% (::ToString(value._dimension, value._type, value._returnType) + (IsMultiSampled(value._dimension) ? to_string(value._numSamples) : ""))
		% value._bindPoint
		% value._bindCount;
	return out;
}